File: fs-regiondeployer/cloudactions/azurefunction/farmer.fs
namespace regiondeployer

open Microsoft.Azure.Management.AppService.Fluent
open regiondeployer
open System.Configuration
open System.Linq

open Microsoft.Azure.Management.Storage.Fluent
open Microsoft.Azure.Management.ResourceManager.Fluent.Core
open System
open System.Collections.Generic
open Microsoft.Azure.Management.Fluent
open Microsoft.Azure.Management.WebSites
open MoonMachine.credentials

type private listlock() = 
    static member val resource:obj = new obj() with get

type farmer() =
    [<Literal>]
    let basefarmername = "mmregionalfarmer"

    [<Literal>]
    let maxfarmernamelength = 40

    member public this.getappsettings() : Dictionary<string, string> =
        let output = new Dictionary<string, string>()
        output.Add("FUNCTIONS_EXTENSION_VERSION", "~2")
        output.Add("AzureWebJobsSecretStorageType", "Files")
        output.Add("WEBSITE_RUN_FROM_PACKAGE", "1")
        output.Add("WEBSITE_NODE_DEFAULT_VERSION", "8.11.1")
        output
        

    member public this.getfunctionapps(logmessage:string -> unit): IEnumerable<IFunctionApp> = 
        let gateway = createauthenticatedgateway()

        lock listlock.resource (fun _ ->        
            gateway.AppServices
                .FunctionApps
                .List()
        ) 

    ///must be executed synchronously
    member public this.getrandomfarmername(logmessage:string -> unit) : string =
        logmessage "getting random farmer name..."
        let randomname = createrandomname(basefarmername, maxfarmernamelength)
        let functionapps = getfunctionapps(logmessage)

        let namechecked = 
            functionapps.Where(fun app -> app.Name = randomname)
                .FirstOrDefault()

        String.concat valueseparator ["checking name is unique"; randomname]
        |> logmessage

        if namechecked = null then
            logmessage "name is unique"
            randomname

        else
            logmessage "name is not unique. recursing..."
            getrandomfarmername logmessage

    member private this.converthostnametoip (hostname:string) : string = //dns lookup using an external api
        let globalconfig = ProjectCredentials.get.matchingcurrentenvironment()

        let accesskey = globalconfig.IpStackKey
        let url = String.concat String.Empty ["http://api.ipstack.com/"; hostname; accesskeyparamkey; accesskey]
        let response = ipstack.Load(url)
        response.Ip

    member public this.externalipalreadyexistsingroup (hostnametotest:string, resourcegroupname:string, logmessage:string -> unit) : bool =  
        let questionappsip = converthostnametoip hostnametotest

        String.concat valueseparator ["checking that farmers external ip "; questionappsip; "does not already exist in resource group"; resourcegroupname]
        |> logmessage

        let groupandhostnamematch (someapp:IFunctionApp) : bool = 
            let queryappsip = converthostnametoip someapp.DefaultHostName
            
            someapp.ResourceGroupName = resourcegroupname && //assuming that function apps cannot have the same external ip unless they exist in the same region.
            questionappsip = queryappsip

        let query =
            getfunctionapps(logmessage)
                .Where(groupandhostnamematch)   
                
        query
        |> fun result -> 
            let existingcount = result.Count()

            match existingcount with
            | 1 -> //assuming there you need an existing function app to get a hostname. 
                String.concat valueseparator ["farmers ip"; questionappsip; "is unique."] 
                |> logmessage           
                false

            | _ -> 
                String.concat valueseparator ["farmers ip"; questionappsip; "already exists"; existingcount.ToString(); "times."]
                logmessage("farmers ip already exists.")
                true

    ///returns a real functionapp. requires existing resourcegroupname.
    ///ensures that any existing farmers, with the same region as the created one, are guaranteed to have different external IP addresses.
    ///created farmer will be on.
    member public this.createfarmer (appname:string, regionname:string, existingresourcegroupname:string, existingstorageaccount:IStorageAccount, logmessage:string -> unit) : IFunctionApp =
        String.concat valueseparator ["creating definition of farmer..."; appname; regionname]
        |> logmessage

        let gateway = createauthenticatedgateway()

        let output = (gateway.AppServices
            .FunctionApps
            .Define(appname)
            .WithRegion(regionname)
            .WithExistingResourceGroup(existingresourcegroupname) //name will be in the format farmername + "group"
            .WithExistingStorageAccount(existingstorageaccount)
            .WithNewConsumptionPlan()
            .WithAppSettings(getappsettings())        
            .WithHttpsOnly(true)
            .WithContainerLoggingEnabled()
            .WithRemoteDebuggingDisabled()
            .Create()) 

        String.concat valueseparator ["definition created"; output.ToString()]
        |> logmessage
        output.Start()

        output

    member public this.deletesinglefarmer(farmerid:string, logmessage:string -> unit) : unit =
        String.concat valueseparator [
            "deleting farmer of id";
            farmerid
        ]
        |> logmessage
        let gateway = createauthenticatedgateway()

        gateway.AppServices
            .FunctionApps
            .DeleteById(farmerid)

        logmessage "farmer deleted."
        ()

    ///deletes any occurrance of a farmer matching input farmer name.
    member public this.deletefarmers (farmername:string, logmessage:string -> unit) : int =    
        let mutable farmersdeleted = 0

        getfunctionapps(logmessage)
            .Where(
                fun farmer -> farmer.Name = farmername
            )
            .Select(
                fun matchingfarmer ->
                    deletesinglefarmer(matchingfarmer.Id, logmessage)
                    farmersdeleted <- (farmersdeleted + 1)
            )
        |> ignore
        farmersdeleted

File: regiondeployer/cloudactions/azurefunction/Farmer.cs
namespace RegionDeployer;

using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Azure.Management.AppService.Fluent;
using RegionDeployer;

public static class ListLock
{
    public static readonly object Resource = new object();
}

public class Farmer
{
    private const string BaseFarmerName = "mmregionalfarmer";
    private const int MaxFarmerNameLength = 40;

    public Dictionary<string, string> GetAppSettings()
    {
        var output = new Dictionary<string, string>
        {
            ["FUNCTIONS_EXTENSION_VERSION"] = "~2",
            ["AzureWebJobsSecretStorageType"] = "Files",
            ["WEBSITE_RUN_FROM_PACKAGE"] = "1",
            ["WEBSITE_NODE_DEFAULT_VERSION"] = "8.11.1"
        };
        return output;
    }

    public IEnumerable<IFunctionApp> GetFunctionApps(Action<string> logMessage, ConfigFile config)
    {
        var authenticator = new AzureAuthenticator();
        var gateway = authenticator.CreateAuthenticatedGateway();
        IEnumerable<IFunctionApp> apps;
        lock (ListLock.Resource)
        {
            apps = gateway.AppServices.FunctionApps.List();
        }
        return apps;
    }

    public string GetRandomFarmerName(Action<string> logMessage)
    {
        logMessage("getting random farmer name...");
        var randomName = CreateRandomName(BaseFarmerName, MaxFarmerNameLength);
        var functionApps = GetFunctionApps(logMessage);

        var nameChecked = functionApps.FirstOrDefault(app => app.Name == randomName);

        logMessage($"checking name is unique: {randomName}");

        if (nameChecked == null)
        {
            logMessage("name is unique");
            return randomName;
        }
        else
        {
            logMessage("name is not unique. recursing...");
            return GetRandomFarmerName(logMessage);
        }
    }

    private string ConvertHostNameToIp(string hostname)
    {
        var globalConfig = ProjectCredentials.Get.MatchingCurrentEnvironment();
        var accessKey = globalConfig.IpStackKey;
        var url = $"http://api.ipstack.com/{hostname}?access_key={accessKey}";
        var response = IpStack.Load(url);
        return response.Ip;
    }

    public bool ExternalIpAlreadyExistsInGroup(string hostnameToTest, string resourceGroupName, Action<string> logMessage)
    {
        var questionAppsIp = ConvertHostNameToIp(hostnameToTest);
        logMessage($"checking that farmer's external IP {questionAppsIp} does not already exist in resource group {resourceGroupName}");

        bool groupAndHostNameMatch(IFunctionApp someApp)
        {
            var queryAppsIp = ConvertHostNameToIp(someApp.DefaultHostName);
            return someApp.ResourceGroupName == resourceGroupName && questionAppsIp == queryAppsIp;
        }

        var query = GetFunctionApps(logMessage).Where(groupAndHostNameMatch);

        var existingCount = query.Count();

        if (existingCount == 1)
        {
            logMessage($"farmer's IP {questionAppsIp} is unique.");
            return false;
        }
        else
        {
            logMessage($"farmer's IP {questionAppsIp} already exists {existingCount} times.");
            return true;
        }
    }

    public IFunctionApp CreateFarmer(string appName, string regionName, string existingResourceGroupName, IStorageAccount existingStorageAccount, Action<string> logMessage)
    {
        logMessage($"creating definition of farmer: {appName}, {regionName}");

        var gateway = CreateAuthenticatedGateway();

        var output = gateway.AppServices.FunctionApps
            .Define(appName)
            .WithRegion(regionName)
            .WithExistingResourceGroup(existingResourceGroupName)
            .WithExistingStorageAccount(existingStorageAccount)
            .WithNewConsumptionPlan()
            .WithAppSettings(GetAppSettings())
            .WithHttpsOnly(true)
            .WithContainerLoggingEnabled()
            .WithRemoteDebuggingDisabled()
            .Create();

        logMessage($"definition created: {output.ToString()}");
        output.Start();

        return output;
    }

    public void DeleteSingleFarmer(string farmerId, Action<string> logMessage)
    {
        logMessage($"deleting farmer of id: {farmerId}");
        var gateway = CreateAuthenticatedGateway();
        gateway.AppServices.FunctionApps.DeleteById(farmerId);
        logMessage("farmer deleted.");
    }

    public int DeleteFarmers(string farmerName, Action<string> logMessage)
    {
        int farmersDeleted = 0;
        var functionApps = GetFunctionApps(logMessage);
        foreach (var farmer in functionApps.Where(farmer => farmer.Name == farmerName))
        {
            DeleteSingleFarmer(farmer.Id, logMessage);
            farmersDeleted++;
        }
        return farmersDeleted;
    }
}